Source/WebCore/css/StyleProperties.cpp

```diff
@@ -575,7 +576,87 @@ static std::optional<CSSValueID> fontStretchKeyword(double value)
+        return isCSSWideKeyword(propertyReferenceKeywordValue(property));
+    };
+
+    auto propertiesResetByShorthandAreExpressible = [&] {
+        if (auto variantCaps = propertyKeywordValue(CSSPropertyFontVariantCaps)) {
+            if (*variantCaps != CSSValueInitial && *variantCaps != CSSValueNormal && *variantCaps != CSSValueSmallCaps)
+                return false;
+        }
+
+        return propertyIsDefault(CSSPropertyFontSizeAdjust, CSSValueNone)
+            && propertyIsDefault(CSSPropertyFontKerning, CSSValueAuto)
+            && propertyIsDefault(CSSPropertyFontVariantLigatures, CSSValueNormal)
+            && propertyIsDefault(CSSPropertyFontVariantPosition, CSSValueNormal)
+            && propertyIsDefault(CSSPropertyFontVariantNumeric, CSSValueNormal)
+            && propertyIsDefault(CSSPropertyFontVariantAlternates, CSSValueNormal)
+            && propertyIsDefault(CSSPropertyFontVariantEastAsian, CSSValueNormal)
+            && propertyIsDefault(CSSPropertyFontFeatureSettings, CSSValueNormal)
+#if ENABLE(VARIATION_FONTS)
+            && propertyIsDefault(CSSPropertyFontOpticalSizing, CSSValueAuto)
+            && propertyIsDefault(CSSPropertyFontVariationSettings, CSSValueNormal)
+#endif
+            && propertyIsDefault(CSSPropertyFontPalette, CSSValueNormal);
+    };
+
+    if (!propertiesResetByShorthandAreExpressible())
+        return emptyString();
+
+    auto commonCSSWideKeyword = [&]() -> std::optional<CSSValueID> {
+        auto keyword = propertyKeywordValue(CSSPropertyFontStyle);
+        if (!keyword || !isCSSWideKeyword(*keyword))
+            return std::nullopt;
+        if (propertyKeywordValue(CSSPropertyFontVariantCaps) != keyword)
+            return std::nullopt;
+        if (propertyKeywordValue(CSSPropertyFontWeight) != keyword)
+            return std::nullopt;
+        if (propertyKeywordValue(CSSPropertyFontStretch) != keyword)
+            return std::nullopt;
+        if (propertyKeywordValue(CSSPropertyFontSize) != keyword)
+            return std::nullopt;
+        if (propertyKeywordValue(CSSPropertyLineHeight) != keyword)
+            return std::nullopt;
+        if (propertyKeywordValue(CSSPropertyFontFamily) != keyword)
+            return std::nullopt;
+        return keyword;
+    };
+
+    auto propertyIsNonInitialCSSWideKeyword = [&](CSSPropertyID property) {
+        auto value = propertyKeywordValue(property);
+        return value && *value != CSSValueInitial && isCSSWideKeyword(*value);
+    };
+
+    auto anyOptionalLonghandIsNonInitialCSSWideKeyword = [&] {
+        return propertyIsNonInitialCSSWideKeyword(CSSPropertyFontStyle)
+            || propertyIsNonInitialCSSWideKeyword(CSSPropertyFontVariantCaps)
+            || propertyIsNonInitialCSSWideKeyword(CSSPropertyFontWeight)
+            || propertyIsNonInitialCSSWideKeyword(CSSPropertyFontStretch)
+            || propertyIsNonInitialCSSWideKeyword(CSSPropertyLineHeight);
+    };
```

Comment:
nit: I tend to define lambdas right before I use them (e.g. before the call to `anyOptionalLonghandIsNonInitialCSSWideKeyword`), reduces a bit of cognitive overhead when reading :)

Reply:
I’ve moved the lambdas around many times while working on this. It turned out the exact sequence of operations was important, and interspersing the lambdas with the code obscured some of the fine points of why things were done in the order they are. That’s why I ended up like this.

But I will change it to follow the "define lambdas right before I use them" pattern; that’s OK with me.